### PR TITLE
fix(eslint-plugin): [no-unnecessary-condition] account for optional chaining on potentially void values

### LINKED ISSUE #6261: Bug: [no-unnecessary-condition] Does not account for optional chaining on potentially-`void` values
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.9.4&sourceType=module&code=DYUwLgBAbghgTgSxgI1BAvBARCgxliGAZwiLEQDsBzCAH2gHsEATAbgFgAoL2RFUAPwA6ANoAGALqsgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK+SpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A

### Repro Code

```typescript
let variable = "abc" as string | void;

variable?.[0];
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

No problems

### Actual Result

The third line reports an error - `Unnecessary optional chain on a non-nullish value.`

### Additional Info

Related to #5766. That PR did not extend to the optional chaining part of this rule.
This is caused by https://github.com/typescript-eslint/typescript-eslint/blob/ccd45d4a998946b7be1161f8c8216bc458e50b4e/packages/type-utils/src/predicates.ts#L28 not checking for `void`, only `null` and `undefined`.

### PR DESCRIPTION
**Pull request by cparros**
## PR Checklist

- [x] Addresses an existing open issue: fixes #6261
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

This PR adds a check for `void` types in the [no-unnecessary-condition] rule. It also adds a valid test case to its valid tests array.

### CODE REVIEW

File: packages/eslint-plugin/tests/rules/no-unnecessary-condition.test.ts [hunk 1]
```diff
@@ -545,6 +545,10 @@ type OptionalFoo = Foo | undefined;
 declare const foo: OptionalFoo;
 foo?.[1]?.length;
     `,
+    `
+let variable = 'abc' as string | void;
+variable?.[0];
+    `,
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/no-unnecessary-condition.test.ts:551]: I looked around the test file and found a few similar tests. Example: `let foo: undefined | { bar: true };`. Could you move this near them?

File: packages/eslint-plugin/src/rules/no-unnecessary-condition.ts [hunk 2]
```diff
@@ -574,10 +574,15 @@ export default createRule<Options, MessageId>({
...
+      const unionParts = unionTypeParts(type);
+      const possiblyVoid = unionParts.some(
+        part => part.flags & ts.TypeFlags.Void,
+      );
       return (
         isTypeAnyType(type) ||
         isTypeUnknownType(type) ||
-        (isNullableType(type, { allowUndefined: true }) && isOwnNullable)
+        (isNullableType(type, { allowUndefined: true }) && isOwnNullable) ||
+        (possiblyVoid && isOwnNullable)
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-unnecessary-condition.ts:585]: The `&& isOwnNullable` is duplicated here. Let's simplify:

```ts
(isOwnNullable && (isNullableType(type, { allowUndefined: true }) || possiblyVoid))
```

File: packages/eslint-plugin/src/rules/no-unnecessary-condition.ts [hunk 3]
```diff
@@ -574,10 +574,15 @@ export default createRule<Options, MessageId>({
         node.type === AST_NODE_TYPES.MemberExpression
           ? !isNullableOriginFromPrev(node)
           : true;
+      const unionParts = unionTypeParts(type);
+      const possiblyVoid = unionParts.some(
+        part => part.flags & ts.TypeFlags.Void,
+      );
       return (
         isTypeAnyType(type) ||
         isTypeUnknownType(type) ||
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-unnecessary-condition.ts:583]: Interesting: `isTypeAnyType(type) and `isTypeUnknownType(type)` both are working fine even if the type is a union. Both of those functions narrow down to checking `isTypeFlagSet(type, ts.TypeFlags.Something)`.

I tried switching from the `possiblyVoid` variable to checking `isTypeFlagSet(type, ts.TypeFlags.Void` - and it seems to work! So I think my hunch about using `unionTypeParts` was wrong 😄.

Tip: TypeScript's flags enums such as `ts.TypeFlags` work with checks like `isTypeFlagSet(type, ts.TypeFlags.One | ts.TypeFlags.Two)`. So you should be able to replace `isTypeAnyType(type) || isTypeUnknownType(type)` with a single `isTypeFlagSet` check.

**Review by JoshuaKGoldberg — DISMISSED** [review]
A good start - glad it's only a little bit of code to fix the bug! Let's get it to _even less_. 🚀

File: packages/eslint-plugin/tests/rules/no-unnecessary-condition.test.ts (hunk 1 above)
Comment (cparros) [reply packages/eslint-plugin/tests/rules/no-unnecessary-condition.test.ts]: Moved!

File: packages/eslint-plugin/src/rules/no-unnecessary-condition.ts (hunk 2 above)
Comment (cparros) [reply packages/eslint-plugin/src/rules/no-unnecessary-condition.ts]: Updated

File: packages/eslint-plugin/src/rules/no-unnecessary-condition.ts (hunk 3 above)
Comment (cparros) [reply packages/eslint-plugin/src/rules/no-unnecessary-condition.ts]: This is great to know! Done.

**Review by bradzacher — APPROVED** [review]
this is looking good to me!
thanks for fixing this!

seemsreasonable

**Pull request merged**